bootloader/efi/efi.go

```diff
@@ -0,0 +1,112 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package efi
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+)
+
+const (
+	defaultEfiVarfsDir    = "/sys/firmware/efivars"
+	defaultEfiVarSysfsDir = "/sys/firmware/efi/vars"
+)
+
+var (
+	readEfiVar  = readEfiVarImpl
+	isSnapdTest = len(os.Args) > 0 && strings.HasSuffix(os.Args[0], ".test")
+)
+
+// ReadEfiVar will attempt to read the binary value of the specified efi
+// variable, specified by it's full name of the variable and vendor ID.
+// It first tries to read from efivarfs wherever that is mounted, and falls back
+// to sysfs if that is unavailable. See
+// https://www.kernel.org/doc/Documentation/filesystems/efivarfs.txt for more
+// details.
+func ReadEfiVar(name string) ([]byte, error) {
+	return readEfiVar(name)
+}
+
+func readEfiVarImpl(name string) ([]byte, error) {
+	// check if we have the efivars fs mounted first, if so then use that
+	// for reading the efi var
+	efiVarDir := filepath.Join(dirs.GlobalRootDir, defaultEfiVarfsDir)
```

Comment:
@pedronis if you think this function is still generic, let me know. I'm inclined to leave it as-is because I already wrote it this way and it's well tested, so :shrug:

Reply:
we are still missing a function to tell whether the system is using EFI or not at all, as we discussed

Reply:
or this should return a clear error about that